meta-balena-common/classes/balena-bootloader.bbclass

```diff
@@ -0,0 +1,642 @@
+        return
+
+    configfilepath = d.getVar("B", True) + '/.config'
+    allSetKernelConfigs = getKernelSetConfigs(configfilepath)
+    configs = d.getVarFlags("BALENA_CONFIGS") or {}
+    firmware_compression = d.getVar('FIRMWARE_COMPRESSION', True)
+
+    if firmware_compression == "1" and \
+        'firmware_compress' not in activatedflags:
+            bb.fatal("Firmware compression is enabled for this device but" \
+                " the kernel does not have support for it")
+
+    for activatedflag in activatedflags:
+        if activatedflag in configs:
+            bb.note("Checking kernel configs for %s." %activatedflag)
+            wantedConfigs = set(configs[activatedflag].split())
+            configured = wantedConfigs.intersection(allSetKernelConfigs)
+            notconfigured = wantedConfigs.difference(configured)
+
+            for config in notconfigured:
+                if not config.endswith('=n'):
+                    bb.warn("Checking for %s in the kernel configs failed for %s."
+                        % (config, activatedflag))
+
+    # Check configs added with resin defconfig
+    resinDefconfig = d.getVar("BALENA_DEFCONFIG_NAME", True)
+    resinDefconfigPath = d.getVar("WORKDIR", True) + '/' +  resinDefconfig
+    wantedConfigs = getKernelSetConfigs(resinDefconfigPath)
+    if wantedConfigs:
+        configured = wantedConfigs.intersection(allSetConfigs)
+        notconfigured = wantedConfigs.difference(configured)
+        for config in notconfigured:
+            if not config.endswith('=n'):
+                bb.warn("Checking for %s in the resin kernel configs failed from %s."
+                    % (config, resinDefconfig))
+}
+addtask kernel_resin_checkconfig after do_kernel_resin_reconfigure before do_compile
+do_kernel_resin_checkconfig[vardeps] += "BALENA_CONFIGS BALENA_CONFIGS_DEPS"
+do_kernel_resin_checkconfig[deptask] += "do_kernel_resin_reconfigure"
+do_kernel_resin_checkconfig[dirs] += "${WORKDIR} ${B}"
+
+do_configure:append () {
+    if [ -f "${DEPLOY_DIR_IMAGE}/balena-keys/kmod.crt" ]; then
+        install -d certs
+        install -m 0655 "${DEPLOY_DIR_IMAGE}/balena-keys/kmod.crt" "certs/"
+    fi
+
+}
+do_configure[depends] += "${@oe.utils.conditional('SIGN_API','','',' balena-keys:do_deploy',d)}"
+# Force compile to depend on the last resin task in the chain
+do_compile[deptask] += "do_kernel_resin_checkconfig"
+# Remove kernel module certificates generated during previous build
+do_configure[cleandirs] += "${@oe.utils.conditional('SIGN_API','','','${B}',d)}"
+do_configure[vardeps] += " \
+    SIGN_API \
+    "
+
+# Because we chain signatures here, the signed artifact is different for each
+# and defined in :prepend for each task
+SIGNING_ARTIFACTS_BASE = "${B}/${KERNEL_OUTPUT_DIR}/${KERNEL_IMAGETYPE}.initramfs"
+addtask sign_efi before do_deploy after do_bundle_initramfs
+addtask sign_gpg before do_deploy after do_sign_efi
```

Comment:
@mtoman question, the chained efi & grub signature is really only needed for EFI systems right? Like the RPI4 will only need the EFI signature. It may not hurt to keep both anyway, just want to know your opinion.

Reply:
That is right, GPG is only there for GRUB. The only DT using this is the `generic-amd64` so once we migrate it to the balena bootloader, we can probably get rid of the `sign-gpg` class altogether.

For the time being & other DTs, it does not hurt to double-sign, we can simply not ship the `.sig` file containing the GPG signature. The only downside is that we need to configure the builders to actually sign with $some GPG key even though the signature won't be used.

Reply:
Actually re-reading my comment the next day it just proves that today we do not need the GPG signing for balena bootloader and we will not need it in the future either so I have removed it.